src/lava/lib/optimization/apps/clustering/solver.py

```diff
@@ -0,0 +1,202 @@
+    can be passed.
+    """
+
+    do_distance_sparsification: bool = False
+    sparsification_algo: str = "cutoff"
+    max_dist_cutoff_fraction: float = 1.0
+    profile_q_mat_gen: bool = False
+    only_gen_q_mat: bool = False
+
+
+@dataclass
+class ClusteringSolution:
+    """Clustering solution holds two dictionaries:
+        - `clustering_id_map` holds a map from cluster center ID to a list
+        of point IDs
+        - `clustering_coords_map` holds a map from the cluster center
+        coordinates to the point coordinates
+    """
+    clustering_id_map: dict = None
+    clustering_coords_map: dict = None
+
+
+class ClusteringSolver:
+    """Solver for clustering problems, given cluster centers.
+    """
+    def __init__(self, clp: ClusteringProblem):
+        self.problem = clp
+        self._solver = None
+        self._profiler = None
+        self.dist_sparsity = 0.
+        self.dist_proxy_sparsity = 0.
+        self.q_gen_time = 0.
+        self.q_shape = None
+        self.raw_solution = None
+        self.solution = ClusteringSolution()
+
+    @property
+    def solver(self):
+        return self._solver
+
+    @property
+    def profiler(self):
+        return self._profiler
+
+    def solve(self, scfg: ClusteringConfig = ClusteringConfig()):
+        """
+        Solve a clustering problem using a given solver configuration.
+
+        Parameters
+        ----------
+        scfg (ClusteringConfig) : Configuration parameters.
+
+        Notes
+        -----
+        The solver object also stores profiling data as its attributes.
+        """
+        # 1. Generate Q matrix for clustering
+        node_list_for_clustering = self.problem.center_coords + \
+            self.problem.point_coords
+        # number of binary variables = total_num_nodes * num_clusters
+        mat_size = len(node_list_for_clustering) * self.problem.num_clusters
+        q_mat_obj = QMatrixClust(
+            node_list_for_clustering,
+            num_clusters=self.problem.num_clusters,
+            lambda_dist=1,
+            lambda_points=100,
+            lambda_centers=100,
+            fixed_pt=True,
+            fixed_pt_range=(-128, 127),
+            clust_dist_sparse_params={
+                "do_sparse": scfg.do_distance_sparsification,
+                "algo": scfg.sparsification_algo,
+                "max_dist_cutoff_fraction": scfg.max_dist_cutoff_fraction},
+            profile_mat_gen=scfg.profile_q_mat_gen)
+        q_mat = q_mat_obj.matrix.astype(int)
+        self.dist_sparsity = q_mat_obj.dist_sparsity
+        self.dist_proxy_sparsity = q_mat_obj.dist_proxy_sparsity
+        if scfg.profile_q_mat_gen:
+            self.q_gen_time = q_mat_obj.time_to_gen_mat
+            self.q_shape = q_mat.shape
+        # 2. Call Lava QUBO solvers
+        if not scfg.only_gen_q_mat:
+            prob = QUBO(q=q_mat)
+            self._solver = OptimizationSolver(problem=prob)
+            hparams = {
```

Comment:
Are these hyperparams always fixed for a clustering problem? I suppose the user doesn't have a chance to tweak them since it's beyond their scope?

Reply:
User can provide custom hyperparameters as an attribute of `SolverConfig`. Only if `scfg.hyperparameters` is an empty dictionary, the default `hparams` will be used to replace it.